packages/core-api/src/apis/implementations/auth/auth0/Auth0Auth.ts

```diff
@@ -0,0 +1,167 @@
+ */
+
+import Auth0Icon from '@material-ui/icons/AcUnit';
+import { DefaultAuthConnector } from '../../../../lib/AuthConnector';
+import { Auth0Session } from './types';
+import {
+  OAuthApi,
+  OpenIdConnectApi,
+  ProfileInfoApi,
+  ProfileInfo,
+  SessionStateApi,
+  SessionState,
+  BackstageIdentityApi,
+  AuthRequestOptions,
+  BackstageIdentity,
+} from '../../../definitions/auth';
+import { OAuthRequestApi, AuthProvider } from '../../../definitions';
+import { SessionManager } from '../../../../lib/AuthSessionManager/types';
+import { RefreshingAuthSessionManager } from '../../../../lib/AuthSessionManager';
+import { Observable } from '../../../../types';
+
+type CreateOptions = {
+  // TODO(Following the words of Rugvip): These two should be grabbed from global config when available, they're not unique to Auth0Auth
+  apiOrigin: string;
+  basePath: string;
+
+  oauthRequestApi: OAuthRequestApi;
+
+  environment?: string;
+  provider?: AuthProvider & { id: string };
+};
+
+export type Auth0AuthResponse = {
+  providerInfo: {
+    accessToken: string;
+    idToken: string;
+    scope: string;
+    expiresInSeconds: number;
+  };
+  profile: ProfileInfo;
+  backstageIdentity: BackstageIdentity;
+};
+
+const DEFAULT_PROVIDER = {
+  id: 'auth0',
+  title: 'Auth0',
+  icon: Auth0Icon,
+};
+
+class Auth0Auth
+  implements
+    OAuthApi,
+    OpenIdConnectApi,
+    ProfileInfoApi,
+    BackstageIdentityApi,
+    SessionStateApi {
+  static create({
+    apiOrigin,
+    basePath,
+    environment = 'development',
+    provider = DEFAULT_PROVIDER,
+    oauthRequestApi,
+  }: CreateOptions) {
+    const connector = new DefaultAuthConnector({
+      apiOrigin,
+      basePath,
+      environment,
+      provider,
+      oauthRequestApi: oauthRequestApi,
+      sessionTransform(res: Auth0AuthResponse): Auth0Session {
+        return {
+          ...res,
+          providerInfo: {
+            idToken: res.providerInfo.idToken,
+            accessToken: res.providerInfo.accessToken,
+            scopes: Auth0Auth.normalizeScopes(res.providerInfo.scope),
+            expiresAt: new Date(
+              Date.now() + res.providerInfo.expiresInSeconds * 1000,
+            ),
+          },
+        };
+      },
+    });
+
+    const sessionManager = new RefreshingAuthSessionManager({
+      connector,
+      defaultScopes: new Set([
+        'openid',
+        `email`,
+        `profile`,
+      ]),
+      sessionScopes: (session: Auth0Session) => session.providerInfo.scopes,
+      sessionShouldRefresh: (session: Auth0Session) => {
+        const expiresInSec =
+          (session.providerInfo.expiresAt.getTime() - Date.now()) / 1000;
+        return expiresInSec < 60 * 5;
+      },
+    });
+
+    return new Auth0Auth(sessionManager);
+  }
+
+  sessionState$(): Observable<SessionState> {
+    return this.sessionManager.sessionState$();
+  }
+
+  constructor(private readonly sessionManager: SessionManager<Auth0Session>) {}
+
+  async getAccessToken(
```

Comment:
Does it actually make sense to implement OAuth for auth0? Do they have their own APIs that you'd want to talk to? Otherwise I'm thinking we just stick to using it as a IdP only

Reply:
I don't have that use case. It was more for someone to use if they wished so.

What are the bare minimum for the Idp? I would need `OpenIdConnectApi` but `ProfileInfoApi, BackstageIdentityApi, SessionStateApi` as well?

Reply:
Bare minimum for providing a Backstage Identity for now would be `ProfileInfoApi`, `BackstageIdentityApi`, `SessionStateApi`. I think it might make sense to go with only that tbh, as afaik the access and ID tokens would be from the the actual provider used to log in to auth0, and at that point we likely want to stick to the purpose-built APIs we already have for that, e.g. googleAuthApi etc.

Reply:
@Rugvip `UserSettings`'s `OIDCProviderSettings` requires for the API Ref to extend `OpenIdConnectApi`. I think we should support at least OpenID/ID token since we might not have a dedicated API for some Idp that uses Auth0 supports.

If you prefer to remove it anyway, does it make sense to have it as a SignIn provider? I could me mixing up the concepts of `Backstage identity` and `Open ID`

Reply:
Removed `OAuthAPI.getaccessToken()` in the meantime

Reply:
The `OIDCProviderSettings` isn't part of the sign-in flow though, that's for OAuth, i.e. access delegation. A bit more info that here: https://github.com/spotify/backstage/blob/master/docs/auth/index.md#accessing-third-party-services. 

`Backstage Identity` isn't fully settled yet, right now it's mostly there so that we can add some constraints to the implementation of auth providers that want to be used for sign-in. Either way the Backstage identity is a partial implementation of Open ID Connect. Looking at for example the Google provider, which implements both Open ID and Backstage Identity, you get two different ID Tokens, one that proves your Google identity, and one that proves your Backstage identity.